src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowJob.java

```diff
@@ -663,6 +663,11 @@ public void replaceAction(@NonNull Action a) {
     @Override protected void performDelete() throws IOException, InterruptedException {
         setDisabled(true);
         Jenkins.get().getQueue().cancel(this);
+
+        for (WorkflowRun workflowRun : getBuilds()) {
```

Comment:
The trouble is that this will make performance worse for those not using `artifact-manager-s3` (or those using it in its default and only supported mode, which never deletes artifacts), since loading potentially thousands of `WorkflowRun` objects from so many little XML files is expensive. Anyway deleting those S3 directories one by one would be inefficient.

Fixing this properly would have to take a different approach I think. https://javadoc.jenkins.io/jenkins/model/ArtifactManager.html#delete() cannot handle this since the whole object is scoped per build. https://javadoc.jenkins.io/jenkins/model/ArtifactManagerFactory.html could have a `delete(Job)` method, perhaps. Or `JCloudsArtifactManager` (or something else in the plugin) could check `ItemListener.onDeleted` and delete the whole directory with one API call.